Check for "add to cart"

DavidCuentasMar opened the pull request:
fixes: #203 
yt link: https://www.youtube.com/watch?v=ewbDOXursUY

dragonfly91 commented on lib/screens/customer_product_detail_page.dart:
```diff
@@ -68,6 +61,38 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
+      color: paletteRoseColor,
+      height: 75,
+      child: Text(
+        '色］［サイズ］を選択してください',
```
You can use this text:

when stockType = sizeAndColor, "色・サイズ　を選択してください"
when stockType = sizeOnly, "サイズを選択してください"
when stockType = colorOnly, "色を選択してください"

DavidCuentasMar replied:
How the quantityOnly should behave? I found this case and not sure how to proceed

dragonfly91 replied:
@DavidCuentasMar we don't need to show any text when stockType = quantityOnly 
These Japanese texts basically are alerting the user "please select size and / or color". This is not needed when only quantity is present. Does this answer your question?

dragonfly91 commented on lib/screens/customer_product_detail_page.dart:
```diff
@@ -312,17 +362,23 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
-        .collection(constants.DBCollections.users)
-        .document(currentSignedInUser.email)
-        .updateData(customer.toMap());
+    if (_selectedItemColor != null && _selectedItemSize != null) {
```
I think this check is valid only for `stockType` = `sizeAndColor`. 

You can add a helper method like this:
```
bool _canAddToCart() {
  if (_product.stock.stockType == StockType.sizeAndColor) {
    return _selectedItemColor != null && _selectedItemSize != null;
  } else if (_product.stock.stockType == StockType.sizeOnly) {
    return _selectedItemSize != null;
  } else if (_product.stock.stockType == StockType.colorOnly) {
    return _selectedItemColor != null;
  }
  return true;
}
```
Then, you can use this method in the `GestureDetector` widget of `_buildAddToCartButton`:
```
if (_canAddToCart()) {
  _addToCart();
} else {
  setState(() {
    _showAddToCartFailedMessage = true;
  });
}
```
This way, you don't have to modify the `_addToCart` method.

dragonfly91 commented on lib/screens/customer_product_detail_page.dart:
```diff
@@ -68,6 +61,59 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
+  }
+
+  Widget _buildAddToCartFailedMessage(StockType productStockType) {
+    String failMessage;
```
A better name could be `errorMessage`

dragonfly91 commented on lib/screens/customer_product_detail_page.dart:
```diff
@@ -266,10 +315,11 @@ class _CustomerProductDetailPageState extends State<CustomerProductDetailPage> {
-          _scaffoldKey.currentState.showSnackBar(
-            _buildAddedToCartNotification(),
-          );
+          if (_selectedItemColor != null && _selectedItemSize != null) {
```
Please refer to my comment below. Snack bar should be shown whenever _addToCart succeeds:
```
if (_canAddToCart()) {
  _addToCart();
  _scaffoldKey.currentState.showSnackBar(      // <------- Add here!
    _buildAddedToCartNotification(),
  );
} else {
  setState(() {
    _showAddToCartFailedMessage = true;
  });
}
```

DavidCuentasMar commented:
In the last commit I removed one streambuilder (it was in _buildAppBar) and moved the other one to build method, this represents only making 1 api call and having access to product object in all widgets. Tell what you think about this.
